test/e2e/storage/vsphere/vsphere.go

```diff
@@ -30,26 +45,238 @@ type VSphere struct {
+		return "", err
+	}
+	taskInfo, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		glog.Errorf("Failed to complete virtual disk creation: %s. err: %+v", vmdkPath, err)
+		return "", err
+	}
+	volumePath := taskInfo.Result.(string)
+	canonicalDiskPath, err := getcanonicalVolumePath(ctx, datacenter, volumePath)
+	if err != nil {
+		return "", err
+	}
+	return canonicalDiskPath, nil
+}
+
+func (vs *VSphere) DeleteVolume(volumePath string, dataCenterRef types.ManagedObjectReference) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	Connect(ctx, vs)
+
+	datacenter := object.NewDatacenter(vs.Client.Client, dataCenterRef)
+	virtualDiskManager := object.NewVirtualDiskManager(datacenter.Client())
+	diskPath := removeStorageClusterORFolderNameFromVDiskPath(volumePath)
+	// Delete virtual disk
+	task, err := virtualDiskManager.DeleteVirtualDisk(ctx, diskPath, datacenter)
+	if err != nil {
+		glog.Errorf("Failed to delete virtual disk. err: %v", err)
+		return err
+	}
+	err = task.Wait(ctx)
+	if err != nil {
+		glog.Errorf("Failed to delete virtual disk. err: %v", err)
+		return err
+	}
+	return nil
+}
+
+// DiskIsAttached returns if disk is attached to the VM using controllers supported by the plugin.
+func (vs *VSphere) DiskIsAttached(volPath string, nodeName string) (bool, error) {
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	Connect(ctx, vs)
+
+	nodeInfo := TestContext.NodeMapper.GetNodeInfo(nodeName)
+	vm := object.NewVirtualMachine(nodeInfo.VSphere.Client.Client, nodeInfo.VirtualMachineRef)
+	volPath = removeStorageClusterORFolderNameFromVDiskPath(volPath)
+	attached, err := isDiskAttached(ctx, vm, volPath)
+	if err != nil {
+		glog.Errorf("DiskIsAttached failed to determine whether disk %q is still attached on node %q",
+			volPath,
+			nodeName)
+	}
+	glog.V(4).Infof("DiskIsAttached result: %q and error: %q, for volume: %q", attached, err, volPath)
+	return attached, err
+}
+
+func (vs *VSphere) IsVMPresent(vmName string, dataCenterRef types.ManagedObjectReference) (isVMPresent bool, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	Connect(ctx, vs)
+	folderMor, err := vs.GetFolderByPath(ctx, dataCenterRef, vs.Config.Folder)
+	if err != nil {
+		return
+	}
+	vmFolder := object.NewFolder(vs.Client.Client, folderMor)
+	vmFoldersChildren, err := vmFolder.Children(ctx)
+	if err != nil {
+		glog.Errorf("Failed to get children from Folder: %s. err: %+v", vmFolder.InventoryPath, err)
+		return
+	}
+	for _, vmFoldersChild := range vmFoldersChildren {
+		if vmFoldersChild.Reference().Type == VirtualMachineType {
+			if object.NewVirtualMachine(vs.Client.Client, vmFoldersChild.Reference()).Name() == vmName {
+				return true, nil
+			}
+		}
+	}
+	return
+}
+
+func (vs *VSphere) DisksAreAttached(nm *NodeMapper, nodeVolumes map[string][]string) (nodeVolumesAttachMap map[string]map[string]bool, err error) {
```

Comment:
There is not need to pass NodeMapper as an argument since TestContext is available globally. Accessing TestContext creates a uniform pattern. I see that NodeManager is used just to further pass it down to convertVolPathsToDevicePaths, I suggest we use TestContext in convertVolPathsToDevicePaths directly.

Reply:
Done.